Source/Details/ASCollectionLayoutState.m

```diff
@@ -20,16 +20,37 @@
 #import <AsyncDisplayKit/ASAssert.h>
 #import <AsyncDisplayKit/ASCellNode+Internal.h>
 #import <AsyncDisplayKit/ASCollectionElement.h>
+#import <AsyncDisplayKit/ASCollectionLayoutContext.h>
 #import <AsyncDisplayKit/ASElementMap.h>
 #import <AsyncDisplayKit/ASLayout.h>
+#import <AsyncDisplayKit/ASPageTable.h>
 
-@implementation ASCollectionLayoutState
+@implementation NSMapTable (ASCollectionLayoutConvenience)
 
-- (instancetype)initWithElements:(ASElementMap *)elements layout:(ASLayout *)layout
++ (NSMapTable<ASCollectionElement *,UICollectionViewLayoutAttributes *> *)elementToLayoutAttributesTable
 {
-  NSMapTable<ASCollectionElement *, UICollectionViewLayoutAttributes *> *attrsMap = [NSMapTable mapTableWithKeyOptions:(NSMapTableObjectPointerPersonality | NSMapTableWeakMemory) valueOptions:NSMapTableStrongMemory];
+  return [NSMapTable mapTableWithKeyOptions:(NSMapTableWeakMemory | NSMapTableObjectPointerPersonality) valueOptions:NSMapTableStrongMemory];
+}
+
+@end
+
+@implementation ASCollectionLayoutState {
+  NSMapTable<ASCollectionElement *,UICollectionViewLayoutAttributes *> *_elementToLayoutAttributesTable;
+  ASPageTable<id, NSMutableSet<UICollectionViewLayoutAttributes *> *> *_pageToLayoutAttributesTable;
```

Comment:
Let's have the value-type here be array rather than set. That way we avoid hashing, equality, and converting-to-array.